src/main/java/radiounit/NokiaLteRadioUnitReceiver.java

```diff
@@ -13,60 +15,185 @@
  */
 public class NokiaLteRadioUnitReceiver implements RadioUnitReceiver {
 
-	private List<Carrier> carriers;
+	private static final int CARRIER_NOT_FOUND_INDEX = -1;
+
+	/*
+	 * The list of carriers is an important part of the radio
+	 * 
+	 * The "volatile" keyword ensures that this resource is not cached, and is
+	 * retrieved directly from memory every time it is accessed.
+	 * 
+	 * The three "-Internal" methods below are all synchronized and have exclusive
+	 * access to this resource.
+	 */
+	private volatile List<Carrier> carriers;
```

Comment:
Just out of curiosity, was there a reason you decided to go with a List rather than a HashMap? I was thinking you could key by carrierId to make retrieval quick and easy, but then I suppose you would also end up storing the Carrier ID twice (once as the key, and once within the Carrier itself). So I guess there are pros and cons to both approaches.

Reply:
Hmm, no I don't think so. I think it was just cause I was aware of ArrayLists, and was less familiar with the advantages of the HashMap. 

By using a HashMap, would that remove the need to iterate through the list in order to get the index of a carrier?

Reply:
That's right. So for example you would put a key/value pair into the hashmap for a carrier, something like 
`carriers.put(ruCarrier.getCarrierId(), ruCarrier);`
And then to retrieve it, you can just call get() on the hashmap and supply the key (in the example above, the carrierId) and it will retrieve the value (the carrier).
`carriers.get(1) // Returns the Carrier that was put into the map with carrierId 1`

Reply:
I like the idea of a HashMap - I think the duplication of data shouldn't be an issue

Reply:
Ah, okay. 

@dLeung09  I'm curious what you think of these suggestions. Since the design we have for these executors/receivers is duplicated across the four sets of classes, which design would you prefer? 

I assume it would be good to maintain a similar approach for all four (though I think I already made some small modifications when compared to your EricssonLte version).